### PR TITLE
Enhance graphical rendering of robot benchmark worlds

### PR DESCRIPTION
**Pull request by fabienrohrer**
## Aim

Improve robotbenchmark benchmarks rendering in order to:

- drop regressions since the last update.
- improve slightly the rendering for obvious issues when it's simple and quick.

## Benchmark vs rendering engine comparative table

*Reflects the current status.*

| Webots r2020a | threejs | Webots R2018b | X3DOM |
| --- | --- | --- | --- |
| highway_driving_1_r2020a | highway_driving_1_three | highway_driving_1_r2018b | highway_driving_1_x3dom |
| highway_driving_2_r2020a | highway_driving_2_three | highway_driving_2_r2018b | highway_driving_2_x3dom |
| humanoid_marathon_1_r2020a | humanoid_marathon_1_threejs | humanoid_marathon_1_r2018b | humanoid_marathon_1_x3dom |
| humanoid_spring_1_r2020a | humanoid_spring_1_threejs | humanoid_spring_1_r2018b | humanoid_spring_1_x3dom |
| humanoid_sprint_2_r2020a | humanoid_sprint_2_threejs | humanoid_sprint_2_r2018b | humanoid_sprint_2_x3dom |
| inverted_pendulum_1_r2020a | inverted_pendulum_1_threejs | inverted_pendulum_1_r2018b | inverted_pendulum_1_x3dom |
| maze_runner_1_r2020a | maze_runner_1_threejs | maze_runner_1_r2018b | maze_runner_1_x3dom |
| obstacle_avoidance_1_r2020a | obstacle_avoidance_1_threejs | obstacle_avoidance_1_r2018b | obstacle_avoidance_1_x3dom |
| pick_and_place_1_r2020a | pick_and_place_1_threejs | pick_and_place_1_r2018b | pick_and_place_1_x3dom |
| pick_and_place_2_threejs | pick_and_place_2_r2020a | pick_and_place_2_r2018b | pick_and_place_2_x3dom |
| pit_escape_1_r2020a | pit_escape_1_threejs | pit_escape_1_r2018b | pit_escape_1_x3dom |
| robot_programming_1_r2020a | robot_programming_1_threejs | robot_programming_1_r2018b | robot_programming_1_x3dom |
| square_path_1_r2020a | square_path_1_threejs | square_path_1_r2018b | square_path_1_x3dom |
| viewpoint_control_1_r2020a | viewpoint_control_1_threejs | viewpoint_control_1_r2018b | viewpoint_control_1_x3dom |
| visual_tracking_1_r2020a | visual_tracking_1_threejs | visual_tracking_1_r2018b | visual_tracking_1_x3dom |
| wall_following_1_r2020a | wall_following_1_threejs | wall_following_1_r2018b | wall_following_1_x3dom |

[Download all the screenshots](https://github.com/omichel/webots/files/3447439/screenshots.zip)

## Global comments

- The light has been changed in every worlds mainly because of the migration to PBR and textured backgrounds.
- Objects in which we worked the PBR appearances (robots, roads, etc.) are already looking much better.
- Several materials would deserve to be upgraded to a true PBR version (with roughness maps for example), particularly some grounds.
- The main rendering difference between Webots r2020a and threejs are already logged as issues.
- `maze_runner` is a special case, because it is hidden.
- `visual_tracking` is a special case, because any rendering modification will affect the controllers. The simplest is certainly to close this benchmark and reopen another visual tracking benchmark (in another PR).
- In some simulation, there is a diff between Webots and threejs rendering (mainly due to #743). This is particularly visible in `pit_escape.wbt` => I focused on the threejs rendering (which is much more visible by the users).

---

- [x] Log

## `webots.min.js` bugs

- ~~when selecting the `humanoid_marathon` ground, all the boundingObjects are selected~~ => it behaves as in Webots, the X3Dom version was buggy.
- ~~texture filtering seems to give weird results~~=> this seems to be a limitation of the browser to support high anisotropy.

## Rendering regressions per benchmarks worlds

*Compare the Webots r2020a/threejs and Webots r2018b/X3DOM.*

- [x] `highway_driving`
    - [x] the overall light seems more tern more colorless than before (the midday effect disappeared) => setup a better light environment
    - [x] fog effect is too much present => increase the fog distance
- [x] `humanoid_marathon`
    - [x] ~~road yellow lines disappeared~~ => the roads PROTO does not allow this anymore, and it's not very important IMO
    - [x] shadows are not visible => Restore shadows
- [x] `humanoid_sprint`
    - [x] Spot light has been replaced by directional light, the shadows and the time counter are not looking good because of this => restore a spot light.
- [x] `inverted_pendulum`
    - [x] Spot lights has been replaced by directional light, the shadows are not looking good because of this => restore a spot lights.
- [x] `maze_runner`
- [x] `obstacle_avoidance`
- [x] `pick_and_place`
    - [x] Spot lights has been replaced by directional light, the shadows are not looking good because of this => restore a spot lights.
- [x] `pit_escape`
    - [x] Light rendering has changed a lot to produce a very lighted world => find a better light tradeoff between light and dark.
    - [x] Spot lights has been replaced by directional light, the shadows are not looking good because of this => restore a spot lights.
- [x] `robot_programming`
- [x] `square_path`
    - Metalness on the walls cause weird reflectivity in threejs
- [x] `viewpoint_control`
- [x] `wall_following`

## Easy improvements per benchmarks worlds

- [x] `highway_driving`
    - [x] grass could be improved with roughness and normal maps => improve the PBR appearance
    - [x] fog color should match with the background to be realist => improve the Fog color
    - [x] there is no reflection on the car glasses because the `mountain` background has a monochrome color => switch to another background with clouds
- [x] `humanoid_marathon`
    - [x] ground looks bad => Use DryMud appearance
    - [x] hydrants looks bad => Use their default appearances
- [x] `humanoid_sprint`
    - [x] grass could be improved with roughness and normal maps => improve the PBR appearance
    - [x] the horizon looks bad => setup a round area instead of rectangular one
    - [x] there is no reflection on the nao plastic parts because the `mountain` background has a monochrome color => switch to another background with clouds
    - [x] improve ground => better texture scale (removed the grass part, removed duplicated plane bounding object, added normal and roughness maps) 
- [x] `inverted_pendulum`
    - [x] This indoor world would deserve another textured background => add another textured background.
    - [x] The original light of this world is very dark => increase the lights intensities.
    - [x] The texture transform of the walls are stretched => fix this.
- [x] `maze_runner`: colors has changed, but for good.
- [x] `obstacle_avoidance`
- [x] `pick_and_place`
    - [x] This factory world would deserve another textured background => add another textured background.
    - [x] The conveyor belt boxes and logo could be easily improved using other appearances like BrushedAluminum, etc. => use another predefined textures.
- [x] `pit_escape`
    - [x] Use SandyGround
- [x] `robot_programming`: colors has changed, but for good.
- [x] `square_path`: colors has changed, but for good.
- [x] `viewpoint_control`: colors has changed, but for good.
- [x] `wall_following`: colors has changed, but for good.

## Webots Assets Modifications

- [x] DryMud
    - [x] DryMud.proto
    - [x] Doc / Icon / Log
- [x] Modification of existing PROTOs
    - [x] TexturedBackgroundLight
        - [x] Add a `luminosity` parameter to echo the `TexturedBackground.luminosity` field.
        - [x] doc / log

### CODE REVIEW

**Comment by fabienrohrer** [comment]
## Results

| Before (R2018b/X3DOM) | This PR (R2019a/threejs) |
| --- | --- |
| highway_driving_2_x3dom | highway_driving |
|  humanoid_marathon_1_x3dom | humanoid_marathon |
| humanoid_spring_1_x3dom | Capture d’écran 2019-08-05 à 15 07 01 |
| inverted_pendulum_1_x3dom | Capture d’écran 2019-08-05 à 17 22 46 |
| pick_and_place_1_x3dom | Capture d’écran 2019-08-05 à 17 12 00 |
|  obstacle_avoidance_1_x3dom | Capture d’écran 2019-08-05 à 17 25 03 |
| square_path_1_x3dom | Capture d’écran 2019-08-05 à 17 41 18 |
| pit_escape_1_x3dom | Capture d’écran 2019-08-06 à 09 42 10 |

**Comment by fabienrohrer** [comment]
👍

**Comment by DavidMansolino** [comment]
:eyes:

File: projects/samples/robotbenchmark/humanoid_sprint/worlds/humanoid_sprint.wbt [hunk 1]
```diff
@@ -27,37 +27,67 @@ Viewpoint {
...
 DEF TRACK Floor {
-  size 11.78 3
-  tileSize 11.78 3
+  translation 0 0 0.39
+  size 11 3
+  tileSize 11 3
   appearance PBRAppearance {
     baseColorMap ImageTexture {
       url [
         "textures/nao_10m_track.jpg"
```
Comment (DavidMansolino) [inline projects/samples/robotbenchmark/humanoid_sprint/worlds/humanoid_sprint.wbt]:
```suggestion
         "textures/nao_10m_track_base_color.jpg"
```

File: projects/samples/robotbenchmark/inverted_pendulum/worlds/inverted_pendulum.wbt [hunk 2]
```diff
@@ -18,8 +18,38 @@ Viewpoint {
   follow "e-puck"
 }
 TexturedBackground {
-}
-TexturedBackgroundLight {
+  texture "empty_office"
```
Comment (DavidMansolino) [inline projects/samples/robotbenchmark/inverted_pendulum/worlds/inverted_pendulum.wbt]: We could add a background of an indoor environment such as this one:
https://hdrihaven.com/hdri/?c=indoor&h=entrance_hall

File: projects/samples/robotbenchmark/pit_escape/worlds/pit_escape.wbt [hunk 3]
```diff
@@ -13,25 +13,29 @@ Viewpoint {
 TexturedBackground {
```
Comment (DavidMansolino) [inline projects/samples/robotbenchmark/pit_escape/worlds/pit_escape.wbt]: We could improve the viewpoint location and orientation to face the sun, the rendering is much nicer.

File: projects/samples/robotbenchmark/pit_escape/worlds/pit_escape.wbt [hunk 4]
```diff
@@ -13,25 +13,29 @@ Viewpoint {
...
 }
-TexturedBackgroundLight {
+Fog {
+  color 0.846693 0.425879 0.0257725
+  fogType "EXPONENTIAL2"
+  visibilityRange 250
+}
+SpotLight {
+  beamWidth 0.7
+  color 0.989548 0.439216 0.032105
```
Comment (DavidMansolino) [inline projects/samples/robotbenchmark/pit_escape/worlds/pit_escape.wbt]: The attenuation is not correct here as it produces the warning:
> WARNING: SpotLight: A quadratic 'attenuation' should be preferred to have a realistic simulation of light. Only the third component of the 'attenuation' field should be greater than 0.

File: projects/samples/robotbenchmark/highway_driving/worlds/highway_driving.wbt [hunk 5]
```diff
@@ -261,6 +264,12 @@ DEF WEBOTS_VEHICLE0 LincolnMKZ {
     }
   ]
   sensorsSlotCenter [
+    SpotLight {
+      beamWidth 0.7
+      direction 0.3 -1 0
+      location -20 40 0
+      castShadows TRUE
+    }
```
Comment (DavidMansolino) [inline projects/samples/robotbenchmark/highway_driving/worlds/highway_driving.wbt]: The attenuation is not correct here as it produces the warning:
> WARNING: SpotLight: A quadratic 'attenuation' should be preferred to have a realistic simulation of light. Only the third component of the 'attenuation' field should be greater than 0.

File: projects/samples/robotbenchmark/humanoid_marathon/worlds/humanoid_marathon.wbt [hunk 6]
```diff
@@ -10,6 +10,15 @@ Viewpoint {
...
 TexturedBackgroundLight {
+  luminosity 0.8
+  castShadows FALSE
+}
+SpotLight {
+  beamWidth 0.7
+  direction 0 -1 0
+  intensity 2
+  location 7 17 -4
+  castShadows TRUE
```
Comment (DavidMansolino) [inline projects/samples/robotbenchmark/humanoid_marathon/worlds/humanoid_marathon.wbt]: The attenuation is not correct here as it produces the warning:
> WARNING: SpotLight: A quadratic 'attenuation' should be preferred to have a realistic simulation of light. Only the third component of the 'attenuation' field should be greater than 0.

File: projects/samples/robotbenchmark/humanoid_sprint/worlds/humanoid_sprint.wbt [hunk 7]
```diff
@@ -27,37 +27,67 @@ Viewpoint {
...
+  luminosity 0.6
+  castShadows FALSE
+}
+SpotLight {
+  beamWidth 0.4
+  cutOffAngle 0.5
+  direction -0.58 -0.8 0.28
+  intensity 1.8
+  location 10 10 -5
+  castShadows TRUE
```
Comment (DavidMansolino) [inline projects/samples/robotbenchmark/humanoid_sprint/worlds/humanoid_sprint.wbt]: The attenuation is not correct here as it produces the warning:
> WARNING: SpotLight: A quadratic 'attenuation' should be preferred to have a realistic simulation of light. Only the third component of the 'attenuation' field should be greater than 0.

File: projects/samples/robotbenchmark/inverted_pendulum/worlds/inverted_pendulum.wbt [hunk 8]
```diff
@@ -18,8 +18,38 @@ Viewpoint {
...
+}
+SpotLight {
+  beamWidth 0.33
+  cutOffAngle 0.56
+  direction 0 -1 0
+  intensity 2
+  location 0 6.5 0.3
+  radius 8
+  castShadows TRUE
```
Comment (DavidMansolino) [inline projects/samples/robotbenchmark/inverted_pendulum/worlds/inverted_pendulum.wbt]: The attenuation is not correct here as it produces the warning:
> WARNING: SpotLight: A quadratic 'attenuation' should be preferred to have a realistic simulation of light. Only the third component of the 'attenuation' field should be greater than 0.

File: projects/samples/robotbenchmark/inverted_pendulum/worlds/inverted_pendulum.wbt [hunk 9]
```diff
@@ -18,8 +18,38 @@ Viewpoint {
...
+}
+SpotLight {
+  beamWidth 0.58
+  cutOffAngle 0.757
+  direction 0 -1 0
+  intensity 2
+  location 11 8 0.3
+  radius 20
+  castShadows TRUE
```
Comment (DavidMansolino) [inline projects/samples/robotbenchmark/inverted_pendulum/worlds/inverted_pendulum.wbt]: The attenuation is not correct here as it produces the warning:
> WARNING: SpotLight: A quadratic 'attenuation' should be preferred to have a realistic simulation of light. Only the third component of the 'attenuation' field should be greater than 0.

File: projects/samples/robotbenchmark/inverted_pendulum/worlds/inverted_pendulum.wbt [hunk 10]
```diff
@@ -18,8 +18,38 @@ Viewpoint {
...
+}
+SpotLight {
+  beamWidth 0.5
+  cutOffAngle 0.72
+  direction 0 -1 0
+  intensity 2
+  location -8 6 0.3
+  radius 8
+  castShadows TRUE
```
Comment (DavidMansolino) [inline projects/samples/robotbenchmark/inverted_pendulum/worlds/inverted_pendulum.wbt]: The attenuation is not correct here as it produces the warning:
> WARNING: SpotLight: A quadratic 'attenuation' should be preferred to have a realistic simulation of light. Only the third component of the 'attenuation' field should be greater than 0.

File: projects/samples/robotbenchmark/pick_and_place/worlds/pick_and_place.wbt [hunk 11]
```diff
@@ -41,10 +41,21 @@ WorldInfo {
...
 TexturedBackgroundLight {
+  texture "factory"
+  castShadows FALSE
+}
+SpotLight {
+  beamWidth 0.7
+  direction 0 -1 0
+  intensity 2.5
+  location 0 4 0
+  castShadows TRUE
```
Comment (DavidMansolino) [inline projects/samples/robotbenchmark/pick_and_place/worlds/pick_and_place.wbt]: The attenuation is not correct here as it produces the warning:
> WARNING: SpotLight: A quadratic 'attenuation' should be preferred to have a realistic simulation of light. Only the third component of the 'attenuation' field should be greater than 0.

**Comment by fabienrohrer** [comment]
I choose to have linear attenuation, because I observed that the lights in threejs are still not well calibrated. But this is a bad solution I'm trying to improve here.

**Comment by fabienrohrer** [comment]
I think I found a better coefficients in threejs, but this remains a tradeoff.

In Webots or threejs, it's difficult to set shadows using quadratic attenuation when the point/spot lights are far. I had to find tradeoffs there too.

With this branch, I think the graphical regression are removed and all the benchmarks look better.

👍

**Comment by DavidMansolino** [comment]
:eyes:

File: docs/reference/changelog-r2020.md [hunk 12]
```diff
@@ -4,8 +4,11 @@
...
 
   - New Features
-    - Added the `Background.luminosity` field which specifies the light contribution of the [Background](background.md) node.
+    - Added the `Background.luminosity` field which specifies the light contribution of the [Background](background.md) node. Open this field in the `TexturedBackground` and the `TexturedBackgroundLight` PROTO nodes.
+    - Added new appearance: `DryMud`
+    - Added new HDR background: `entrance_hall`
     - Improved the Webots online 3D viewer: `webots.min.js`
       - Improved support of the Webots rendering pipeline: supported the Bloom post-processing effect.
       - Added support for the `ImageTexture.filtering` field.
+    - Improved graphics quality of the [robotbenchmark](https://www.robotbenchmark.net) worlds.
```
Comment (omichel) [inline docs/reference/changelog-r2020.md]: The official URL is https://robotbenchmark.net

**Comment by fabienrohrer** [comment]
I also just talked with Stefania and she told me that square_path and inverted_pendulum worlds are too bright, I'm fixing this too.

**Comment by fabienrohrer** [comment]
👍

**Comment by stefaniapedrazzi** [comment]
:eyes:

**Pull request merged**